### ISSUE TITLE
Allow functions and "registered" custom types for the field type

### ISSUE THREAD

**Issue by serkandurusoy** [aldeed/meteor-simple-schema#523]
Continuing on the discussion from the [forum thread](https://forums.meteor.com/t/are-match-and-check-here-to-stay-is-a-new-alternative-in-the-works/15211/27?u=serkandurusoy)

@aldeed
> Yeah, I'm happy to look into that, but it would be useful to have a couple realistic real world use cases in mind to understand what the API needs to entail.

There are endless possibilities here, first of all regex would become redundant and libraries like isjs can be used to validate data types.

Furthermore, app specific checks could be introduced

```javascript
SimpleSchema.registerType({
  name: "dictionaryWord",
  constraint: function() {
    var isString = _.isString(this.value);
    var existsInDictionary: isString && Dictionary.findOne({word: this.value});
    return !!existsInDictionary
  },
  message: '[value] is not a valid dictionary word for [label]'
});

BlogArticle = new SimpleSchema({
  title: {type: String},
  body: {type: String},
  tag: {type: existsInDictionary}
})
```

I know these could very well be introduced as functions and called from the `custom` definitions, but still something similar to this would make the API more pleasing and consistent. It would also defeat the arguments that SS is not as extensible as check.

**Comment by aldeed** [comment]
@serkandurusoy This can be done right now, so is it correct to say you're just suggesting a different API? I agree your API idea might be a bit easier to understand.

Here is how it would work now:

```js
SimpleSchema.messages({
  notDictionaryWord: '[value] is not a valid dictionary word for [label]'
});

var existsInDictionary = function () {
  var isString = _.isString(this.value);
  var existsInDictionary: isString && Dictionary.findOne({word: this.value});
  if (!existsInDictionary) return 'notDictionaryWord';
};

BlogArticle = new SimpleSchema({
  title: {type: String},
  body: {type: String},
  tag: {type: String, custom: existsInDictionary}
})
```

**Comment by serkandurusoy** [comment]
Yes! I do confirm that `custom` does allow for it and my suggestion is a
different API for the same purpose, one that minimizes the use of `custom`
to truly custom and schema specific instances where type can be more
generalized.

Sent from my iPhone



On Dec 22, 2015, at 20:17, Eric Dobbertin <notifications@github.com> wrote:

@serkandurusoy <https://github.com/serkandurusoy> This can be done right
now, so is it correct to say you're just suggesting a different API? I
agree your API idea might be a bit easier to understand.

Here is how it would work now:

SimpleSchema.messages({
  notDictionaryWord: '[value] is not a valid dictionary word for [label]'
});
var existsInDictionary = function () {
  var isString = _.isString(this.value);
  var existsInDictionary: isString && Dictionary.findOne({word: this.value});
  if (!existsInDictionary) return 'notDictionaryWord';
};

BlogArticle = new SimpleSchema({
  title: {type: String},
  body: {type: String},
  tag: {type: String, custom: existsInDictionary}
})

—
Reply to this email directly or view it on GitHub
<https://github.com/aldeed/meteor-simple-schema/issues/523#issuecomment-166694981>
.
